bundle/libraries/upload.go

```diff
@@ -0,0 +1,226 @@
+package libraries
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/databricks/cli/bundle"
+	"github.com/databricks/cli/libs/cmdio"
+	"github.com/databricks/cli/libs/diag"
+	"github.com/databricks/cli/libs/dyn"
+	"github.com/databricks/cli/libs/filer"
+	"github.com/databricks/cli/libs/log"
+
+	"github.com/databricks/databricks-sdk-go"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// The Files API backend has a rate limit of 10 concurrent
+// requests and 100 QPS. We limit the number of concurrent requests to 5 to
+// avoid hitting the rate limit.
+var maxFilesRequestsInFlight = 5
+
+func Upload() bundle.Mutator {
+	return &upload{}
+}
+
+func UploadWithClient(client filer.Filer) bundle.Mutator {
+	return &upload{
+		client: client,
+	}
+}
+
+type upload struct {
+	client filer.Filer
+}
+
+type configLocation struct {
+	configPath dyn.Path
+	location   dyn.Location
+}
+
+// Collect all libraries from the bundle configuration and their config paths.
+// By this stage all glob references are expanded and we have a list of all libraries that need to be uploaded.
+// We collect them from task libraries, foreach task libraries  environment dependencies and artifacts.
+// We return a map of library source to a list of config paths and locations where the library is used.
+// We use map so we don't upload the same library multiple times.
+// Instead we upload it once and update all the config paths to point to the uploaded location.
+func collectLocalLibraries(b *bundle.Bundle) (map[string][]configLocation, error) {
+	libs := make(map[string]([]configLocation))
+
+	patterns := []dyn.Pattern{
+		taskLibrariesPattern.Append(dyn.AnyIndex(), dyn.Key("whl")),
+		taskLibrariesPattern.Append(dyn.AnyIndex(), dyn.Key("jar")),
+		forEachTaskLibrariesPattern.Append(dyn.AnyIndex(), dyn.Key("whl")),
+		forEachTaskLibrariesPattern.Append(dyn.AnyIndex(), dyn.Key("jar")),
+		envDepsPattern.Append(dyn.AnyIndex()),
+	}
+
+	for _, pattern := range patterns {
+		err := b.Config.Mutate(func(v dyn.Value) (dyn.Value, error) {
+			return dyn.MapByPattern(v, pattern, func(p dyn.Path, v dyn.Value) (dyn.Value, error) {
+				source := v.MustString()
+				if !IsLibraryLocal(source) {
+					return v, nil
+				}
+
+				source = filepath.Join(b.RootPath, source)
+				libs[source] = append(libs[source], configLocation{
+					configPath: p.Append(), // Hack to get the copy of path
+					location:   v.Location(),
+				})
+
+				return v, nil
+			})
+		})
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	artifactPattern := dyn.NewPattern(
+		dyn.Key("artifacts"),
+		dyn.AnyKey(),
+		dyn.Key("files"),
+		dyn.AnyIndex(),
+	)
+
+	err := b.Config.Mutate(func(v dyn.Value) (dyn.Value, error) {
+		return dyn.MapByPattern(v, artifactPattern, func(p dyn.Path, v dyn.Value) (dyn.Value, error) {
+			file := v.MustMap()
```

Comment:
This will panic if the value is nil.

This is invalid of course, but best to be defensive and check that it is the expected type and return if it isn't.